# Spam dataset excercises

## 1.0 Setup

In [12]:
import pandas as pd
import datatable as dt
import numpy as np
print(dt.__version__)

0.8.0


In [27]:
data = pd.read_csv('~/personal/daa_2021s2/Exercises/05 Classification/Datasets/spambase.csv')
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  y  
0                       278  1  
1                      1028  1  
2                      2259  1  
3                       191  1  
4                       191  1  

[5 rows x 58 columns]

## 2.0 create simple predictions
SImple predictions for excercise is classifying all observations with 'char_freq_$' as spam

In [48]:
data['pred_1'] =  np.where(data['char_freq_!']>= 0.045, 1, 0)
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_(  char_freq_[  \
0             0.00            0.00  ...        0.000          0.0   
1             0.00            0.94  ...        0.132          0.0   
2             0.64            0.25  ...        0.143          0.0   
3             0.31            0.63  ...        0.137          0.0   
4             0.31            0.63  ...        0.135          0.0   

   char_freq_!  char_freq_$  char_freq_#  capital_run_length_average  \
0        0.778        0.000        0.000                       3.756   
1        0.372        0.180        0.048                       5.114   
2        0.276        0.184        0.010                       9.821   
3        0.137        0.000        0.000                       3.537   
4        0.135        0.000        0.000                       3.537   

   capital_run_length_longest  capital_run_length_total  y  pred_1  
0                          61                       278  1       1  
1                         101                      1028  1       1  
2                         485                      2259  1       1  
3                          40                       191  1       1  
4                          40                       191  1       1  

[5 rows x 59 columns]

In [47]:
confusion_matrix = pd.crosstab(data['y'], data['pred_1'], rownames=['Actual'], colnames=['Predicted'])
print (confusion_matrix)


Predicted     0     1
Actual               
0          2137   651
1           361  1452


|  | pred 0 | pred 1 | total |
| --- | --- | --- |--- |
| Actual 0| 2,137|651 |2,788|
|Actual 1|361|1,452|1,813|
||2,498|2,103|4601|

misclassification	0.219952184	
Accuracy	0.780047816	
precision	0.690442225	
sensitivity	0.800882515	
specificity	0.766499283	
false neg	0.199117485	
false pos rate	0.233500717	


In [15]:
# Create a list of headings for the data.
namearray = [
'word_freq_make',
'word_freq_address',
'word_freq_all',
'word_freq_3d',
'word_freq_our',
'word_freq_over',
'word_freq_remove',
'word_freq_internet',
'word_freq_order',
'word_freq_mail',
'word_freq_receive',
'word_freq_will',
'word_freq_people',
'word_freq_report',
'word_freq_addresses',
'word_freq_free',
'word_freq_business',
'word_freq_email',
'word_freq_you',
'word_freq_credit',
'word_freq_your',
'word_freq_font',
'word_freq_000',
'word_freq_money',
'word_freq_hp',
'word_freq_hpl',
'word_freq_george',
'word_freq_650',
'word_freq_lab',
'word_freq_labs',
'word_freq_telnet',
'word_freq_857',
'word_freq_data',
'word_freq_415',
'word_freq_85',
'word_freq_technology',
'word_freq_1999',
'word_freq_parts',
'word_freq_pm',
'word_freq_direct',
'word_freq_cs',
'word_freq_meeting',
'word_freq_original',
'word_freq_project',
'word_freq_re',
'word_freq_edu',
'word_freq_table',
'word_freq_conference',
'char_freq_;',
'char_freq_(',
'char_freq_[',
'char_freq_!',
'char_freq_$',
'char_freq_#',
'capital_run_length_average',
'capital_run_length_longest',
'capital_run_length_total',
'Spam_fl' ]

# Read in the data from the Stanford website.
spam = pd.read_csv('~/personal/daa_2021s2/Exercises/05 Classification/Datasets/spam.data', delim_whitespace=True, 
                   header=None,
                   names=namearray
                 )



In [16]:
# Check the dimensions of the data.
print(spam.info())

# Print the first 10 observations from the data.
print(spam.head(10))


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4601 entries, 0 to 4600
Data columns (total 58 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   word_freq_make              4601 non-null   float64
 1   word_freq_address           4601 non-null   float64
 2   word_freq_all               4601 non-null   float64
 3   word_freq_3d                4601 non-null   float64
 4   word_freq_our               4601 non-null   float64
 5   word_freq_over              4601 non-null   float64
 6   word_freq_remove            4601 non-null   float64
 7   word_freq_internet          4601 non-null   float64
 8   word_freq_order             4601 non-null   float64
 9   word_freq_mail              4601 non-null   float64
 10  word_freq_receive           4601 non-null   float64
 11  word_freq_will              4601 non-null   float64
 12  word_freq_people            4601 non-null   float64
 13  word_freq_report            4601 